components/raftstore/src/coprocessor/region_info_accessor.rs

```diff
@@ -563,46 +565,61 @@ impl RegionCollector {
     }
 
     /// Used by the in-memory engine (if enabled.)
-    /// If `count` is 0, return all the regions for which this node is the
-    /// leader. Otherwise, return the top `count` regions from
-    /// `self.region_activity`. Top regions are determined by comparing
-    /// `read_keys + written_keys` in each region's most recent region stat.
-    /// If count > 0 and size of `self.region_activity` is `> 0`, remove but the
-    /// top `count` regions from `self.region_activity`.
+    /// If `count` is 0, return all the regions.
+    ///
+    /// Otherwise, return the top `count` regions for which this node is the
+    /// leader from `self.region_activity`. Top regions are determined by
+    /// comparing `read_keys` in each region's most recent
+    /// region stat.
     ///
     /// Note: this function is `O(N log(N))` with respect to size of
     /// region_activity. This is acceptable, as region_activity is populated
     /// by heartbeats for this node's region, so N cannot be greater than
     /// approximately `300_000``.
-    pub fn handle_get_top_regions(&mut self, count: usize, callback: Callback<Vec<Region>>) {
+    pub fn handle_get_top_regions(&mut self, count: usize, callback: Callback<TopRegions>) {
+        let compare_fn = |a: &RegionActivity, b: &RegionActivity| {
+            let a = a.region_stat.read_keys;
+            let b = b.region_stat.read_keys;
+            b.cmp(&a)
+        };
         let top_regions = if count == 0 {
             self.regions
                 .values()
-                .filter(|ri| ri.role == StateRole::Leader)
-                .map(|ri| ri.region.clone())
+                .map(|ri| {
+                    (
+                        ri.region.clone(),
+                        self.region_activity.get(&ri.region.get_id()),
+                    )
+                })
+                .sorted_by(|(_, a), (_, b)| match (a, b) {
+                    (None, None) => Ordering::Equal,
+                    (None, Some(_)) => Ordering::Greater,
+                    (Some(_), None) => Ordering::Less,
+                    (Some(a), Some(b)) => compare_fn(a, b),
```

Comment:
If there are many regions that have a lot of read requests but without MVCC issue, these regions will be treat as top regions and cached in the in-memory engine. This is a waste for the memory. I think calculate top regions according the MVCC issue severity combine with there is some read traffic for these regions is better.

Reply:
I am going to add a micro-bench on gathering data and computing top N. Then in a follow on PR, I will look at plugging in gathering MVCC stats from RocksDb and using that. Does this make sense? (Unless I am missing a place where MVCC versions are already available, I wish to avoid having exceeding the size of this PR. cc @zhangjinpeng87)

Reply:
sounds good

Reply:
Let's merge the FRAMEWORK first, and then use following PRs to tuning the inside load&evict algorithms.